File: youngplatform_docs/examples/js/api_v4.js
const axios = require('axios');
const crypto = require('crypto');
const jwt = require('jsonwebtoken');

// Configuration variables
const API_URL = 'https://api.youngplatform.com';
const KEY_ID =""  // # Replace with your key ID string
const PUBLIC_KEY =""  // # Replace with your public key string
const PRIVATE_KEY =""  // # Replace with your private key string

/**
 * Generate a JWT token with optional payload hashing.
 * @param {string} publicKey - The public key string.
 * @param {string} privateKey - The private key string.
 * @param {string} [payload] - The payload to include in the JWT hash (optional).
 * @returns {string} - Encoded JWT token.
 */
function generateJwt(publicKey, privateKey, payload = "") {
    const iat = Math.floor(Date.now() / 1000);
    const exp = iat + 30;

    // Hash the payload
    const hashPayload = crypto.createHash('sha256').update(payload).digest('hex');

    // Define JWT claims
    const jwtClaims = {
        "sub": publicKey,
        "iat": iat,
        "exp": exp,
        "hash_payload": hashPayload
    };

    // Generate and return the JWT token
    return jwt.sign(jwtClaims, privateKey, { algorithm: 'HS256' });
}

/**
 * Fetch user balances.
 * @returns {Promise<Object>} - Balance data.
 */
async function getBalance() {
    try {
        const url = `${API_URL}/api/v4/private/balances`;
        const token = generateJwt(PUBLIC_KEY, PRIVATE_KEY);
        const headers = {
            "Authorization": token,
            "X-Api-Key-Id": KEY_ID,
        };
        const response = await axios.get(url, { headers });
        return response.data;
    } catch (error) {
        console.error("Error fetching balances:", error);
        throw error;
    }
}

/**
 * Cancel an order by ID.
 * @param {string} orderId - The ID of the order to cancel.
 * @returns {Promise<Object>} - Response data from the API.
 */
async function cancelOrder(orderId) {
    try {
        const url = `${API_URL}/api/v4/private/cancel`;
        const body = { orderId };
        const bodyString = JSON.stringify(body);
        const token = generateJwt(PUBLIC_KEY, PRIVATE_KEY, bodyString);
        const headers = {
            "Authorization": token,
            "X-Api-Key-Id": KEY_ID,
            "Content-Type": "application/json",
        };
        const response = await axios.post(url, bodyString, { headers });
        return response.data;
    } catch (error) {
        if (error.response && error.response.status === 400) {
            return error.response.data;
        }
        console.error("Error cancelling order:", error);
        throw error;
    }
}

// Main execution flow
(async () => {
    try {
        const balanceData = await getBalance();
        console.log("Balances:", balanceData);

        const cancelResponse = await cancelOrder("1");
        console.log("Cancel Order Response:", cancelResponse);
    } catch (error) {
        console.error("Error in main flow:", error);
    }
})();

File: youngplatform_docs/examples/python/api_v4.py
import time
import hashlib
import jwt
import requests
import json



# Configuration variables
KEY_ID = ""  # Replace with your key ID string
PUBLIC_KEY = ""  # Replace with your public key string
PRIVATE_KEY = ""  # Replace with your private key string
BASE_URL = "https://api.youngplatform.com/api/v4/private"


def generate_jwt(public_key, private_key, payload=None):
    """
    Generate a JWT token with optional payload hashing.

    :param public_key: The public key string.
    :param private_key: The private key string.
    :param payload: The payload to include in the JWT hash (optional).
    :return: Encoded JWT token.
    """
    iat = int(time.time())
    exp = iat + 30

    # Use an empty string if no payload is provided
    payload = payload or ""

    # Hash the payload
    hash_payload = hashlib.sha256(payload.encode('utf-8')).hexdigest()

    # Define JWT claims
    jwt_claims = {
        "sub": public_key,
        "iat": iat,
        "exp": exp,
        "hash_payload": hash_payload,
    }

    # Generate and return the JWT token
    return jwt.api_jwt.encode(jwt_claims, private_key, algorithm="HS256")


def make_request(method, endpoint, jwt_token, payload=None):
    """
    Make an HTTP request to the API.

    :param method: HTTP method (GET or POST).
    :param endpoint: API endpoint to call.
    :param jwt_token: JWT token for authentication.
    :param payload: JSON payload for POST requests (optional).
    :return: Response object.
    """
    headers = {
        "Authorization": jwt_token,
        "X-Api-Key-Id": KEY_ID,
    }
    if payload:
        headers["Content-Type"] = "application/json"

    url = f"{BASE_URL}/{endpoint}"

    # Send the appropriate request
    if method == "GET":
        response = requests.get(url, headers=headers)
    elif method == "POST":
        response = requests.post(url, headers=headers, data=json.dumps(payload))
    else:
        raise ValueError("Unsupported HTTP method")

    # Handle errors
    if response.status_code != 200:
        print(f"Error: {response.status_code}\n{response.text}")
        exit()

    return response


def main():
    # Generate a JWT token for a GET request
    jwt_token = generate_jwt(PUBLIC_KEY, PRIVATE_KEY)

    # GET request for balances
    balances_response = make_request("GET", "balances", jwt_token)
    print(balances_response.json())

    # Payload for cancelling an order
    cancel_order_payload = {
        "orderId": 100
    }

    # Generate a JWT token for a POST request
    jwt_token_body = generate_jwt(PUBLIC_KEY, PRIVATE_KEY, json.dumps(cancel_order_payload))

    # POST request to cancel an order
    cancel_response = make_request("POST", "cancel", jwt_token_body, cancel_order_payload)
    print(cancel_response.json())


if __name__ == "__main__":
    main()
